### DISCUSSION TITLE
rerun onSuccess method of a cached query

### DISCUSSION THREAD

**Discussion by crisfcodes** [TanStack/query#4535] [Q&A]
## What I have

I am getting the user's location (latitude/longitude) which I use to call a google [geocode API](https://developers.google.com/maps/documentation/javascript/examples/geocoding-reverse), unless the user's coords change, the request is not running again, since the query it uses the user's coords as queryKey array dependecy.

## The problem

the problem is that I'm running some operations in the `onSuccess` query method, this method is only run when any of the queryKey dependencies change, and I mentioned this not happen.

**How to run the `onSuccess` method whether the queryKey dependencies change or not?**

## Reference code

```typescript
export const useGoogleReverseGeocoding = (coords) => {
  const url = 'someUrl';
  const request = createClient(); // axios abstraction

  return useQuery({
    queryKey: ['google-geocode', coords],
    queryFn: request,
    enabled: !!coords,
    onSuccess: (data) => {
      // ...operation with data
      // I need to run this whether the queryKey dependencies change or not
    },
  });
};
```

@TkDodo

**Comment by TkDodo** [comment]
1) why do you keep @-mentioning me personally on your questions .. the discussion forum is well moderated ...

2) `onSuccess` is tied to a successful request being made, and it's also called once for every observer. So if you call `useGoogleReverseGeocoding` twice in two components, `onSuccess` will be called for each of them on a successful request. However, if data is read from the cache only, it will not be called.

We might add a different callback `onDataChanged` that will be called every time data is changing, but it's the same as:

```
useEffect(() => {}, [data]
```

because `data` is referentially stable.

So, it really depends on what that "operation" is. I don't see a good use-case for `onSuccess` tbh 🤷
